### PR TITLE
executor: introduce hashRowContainer.GetMatchedRows method

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/36847

Problem Summary:

### What is changed and how it works?
Introduce the hashRowContainer.GetMatchedRows method since we do not need the Ptr info in HashJoinExec.joinMatchedProbeSideRow2Chunk

### Check List

Tests 

Before this commit:
TPCDS Q95 costs 126s

After this commit:
TPCDS Q95 costs 113s
飞书20220803-171505

Side effects

- N/A

Documentation

- N/A

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by guo-shaoge — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

File: executor/hash_table.go [hunk 1]
```diff
@@ -112,10 +112,18 @@ func (c *hashRowContainer) ShallowCopy() *hashRowContainer {
 	return &newHRC
 }
 
+// GetMatchedRows get matched rows from probeRow. It can be called
+// in multiple goroutines while each goroutine should keep its own
+// h and buf.
```
Comment (zanmato1984) [inline executor/hash_table.go:117]: ”h and buf"?

File: executor/join.go [hunk 2]
```diff
@@ -532,10 +532,11 @@ func (e *HashJoinExec) joinMatchedProbeSideRow2ChunkForOuterHashJoin(workerID ui
 	}
 	return true, joinResult
 }
+
 func (e *HashJoinExec) joinMatchedProbeSideRow2Chunk(workerID uint, probeKey uint64, probeSideRow chunk.Row, hCtx *hashContext,
 	rowContainer *hashRowContainer, joinResult *hashjoinWorkerResult) (bool, *hashjoinWorkerResult) {
 	var err error
-	e.buildSideRows[workerID], e.buildSideRowPtrs[workerID], err = rowContainer.GetMatchedRowsAndPtrs(probeKey, probeSideRow, hCtx, e.buildSideRows[workerID], e.buildSideRowPtrs[workerID])
+	e.buildSideRows[workerID], err = rowContainer.GetMatchedRows(probeKey, probeSideRow, hCtx, e.buildSideRows[workerID])
```
Comment (zanmato1984) [inline executor/join.go:539]: How is this change preventing false sharing of `e.buildSideRows`?

File: executor/hash_table.go (hunk 1 above)
Comment (XuHuaiyu) [reply executor/hash_table.go:117]: This is a copy of GetMatchedRowsAndPtrs.

File: executor/join.go (hunk 2 above)
Comment (XuHuaiyu) [reply executor/join.go:539]: This PR does not handle the false sharing thing

**Pull request merged**